File: packages/js-dpp/lib/decodeProtocolEntityFactory.js
const varint = require('varint');
const ProtocolVersionParsingError = require('./errors/consensus/basic/decode/ProtocolVersionParsingError');
const SerializedObjectParsingError = require('./errors/consensus/basic/decode/SerializedObjectParsingError');

const { decode } = require('./util/serializer');

function decodeProtocolEntityFactory() {
  /**
   * @typedef {decodeProtocolEntity}
   * @param {Buffer} buffer
   * @return {[number, Object]}
   */
  function decodeProtocolEntity(buffer) {
    // Parse protocol version from the first 4 bytes
    let protocolVersion;
    try {
      protocolVersion = varint.decode(buffer);
    } catch (error) {
      const consensusError = new ProtocolVersionParsingError(error.message);

      consensusError.setParsingError(error);

      throw consensusError;
    }

    let rawEntity;
    try {
      rawEntity = decode(
        buffer.slice(varint.encodingLength(protocolVersion), buffer.length),
      );
    } catch (error) {
      const consensusError = new SerializedObjectParsingError(error.message);

      consensusError.setParsingError(error);

      throw consensusError;
    }

    return [protocolVersion, rawEntity];
  }

  return decodeProtocolEntity;
}

module.exports = decodeProtocolEntityFactory;

File: packages/rs-dpp/src/encoding/decode_protocol_entity_factory.rs
use platform_value::Value;

use crate::consensus::basic::decode::SerializedObjectParsingError;
use crate::consensus::basic::BasicError;
use crate::serialization_traits::{PlatformDeserializable, ValueConvertible};
use crate::util::deserializer;
use crate::util::deserializer::SplitProtocolVersionOutcome;
use crate::{errors::consensus::ConsensusError, errors::ProtocolError};

#[derive(Default, Clone, Copy)]
pub struct DecodeProtocolEntity {}

impl DecodeProtocolEntity {
    pub fn decode_protocol_entity<T: PlatformDeserializable>(
        buffer: impl AsRef<[u8]>,
    ) -> Result<(u32, T), ProtocolError> {
        let SplitProtocolVersionOutcome {
            protocol_version,
            main_message_bytes: document_bytes,
            ..
        } = deserializer::split_protocol_version(buffer.as_ref())?;

        let protocol_entity = T::deserialize(document_bytes).map_err(|e| {
            ConsensusError::BasicError(BasicError::SerializedObjectParsingError(
                SerializedObjectParsingError::new(e.to_string()),
            ))
        })?;

        //todo deal with version

        Ok((protocol_version, protocol_entity))
    }

    pub fn decode_protocol_entity_to_value<T: PlatformDeserializable + ValueConvertible>(
        buffer: impl AsRef<[u8]>,
    ) -> Result<(u32, Value), ProtocolError> {
        let SplitProtocolVersionOutcome {
            protocol_version,
            main_message_bytes: document_bytes,
            ..
        } = deserializer::split_protocol_version(buffer.as_ref())?;

        let protocol_entity = T::deserialize(document_bytes).map_err(|e| {
            ConsensusError::BasicError(BasicError::SerializedObjectParsingError(
                SerializedObjectParsingError::new(format!("Decode protocol entity: {:#?}", e)),
            ))
        })?;

        Ok((protocol_version, protocol_entity.to_object()?))
    }
}
